### ISSUE TITLE
[no-non-null-assertion] suggests impossible fix when used as L-value in assignment

### ISSUE THREAD

**Issue by thw0rted** [typescript-eslint/typescript-eslint#2910]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
    "extends": [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended"
    ],
}
```

```TS
interface Foo { arr?: string[]; }
const f: Foo = { arr: ["a"] };
f.arr!.length = 0;
```

**Expected Result**

There probably isn't any automated way of suggesting an improvement to this code.  The user could add a guard (`if (f.arr) { ... }`) even though they know it will always be true, or maybe declare it differently like

````ts
const arr = ["a"];
const f: Foo = { arr };
arr.length = 0;
````

but there's no one fix to suggest (other than just using the assertion operator anyway...).

**Actual Result**

The rule suggests "Consider using the optional chaining operator" and offers to change the code to `f.arr?.length = 0;`, which is a compiler error.  (I'm applying the suggestion via the eslint VSCode plugin, but assume it would probably work the same way via CLI, right?)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.11.0` |
| `@typescript-eslint/parser`        | `4.11.0` |
| `TypeScript`                       | `4.1.3` |
| `ESLint`                           | `7.15.0` |

**Comment by thw0rted** [comment]
I'm looking through the [eslint docs about writing rules](https://eslint.org/docs/developer-guide/working-with-rules) and can't seem to figure out if a given rule is allowed to be conditionally "fixable", i.e. if one rule can have a suggestion or not based on the nature of the flagged code.  I also assume it'd be a huge headache to split the rule into one that catches (fixable) r-value usage and one that catches (non-fixable) l-values.

**Comment by bradzacher** [comment]
This rule uses a suggestion fixers.

Automatic fixes should always be safe, and should never cause syntax / type errors. This is because they are intended to be applied by automation (eg the CLI)

Suggestion fixers, OTOH, are not intended to be safe for all cases. This is because they are not able to be applied by automation. They can only be applied manually (and one by one) via an IDE.

It's up to the engineer to read the description, assess its safety, and apply if it's safe.

Many rules have similar suggestions which can cause breakages if applied incorrectly. For example our `no-explicit-any` lint rule has a suggestion fixer to change an `any` to `unknown` - which will cause type errors for must cases.

**Comment by thw0rted** [comment]
Thanks for being patient with all these issues -- I'm still learning the capabilities and limitations of the linter, so the distinction between automatic fixes and suggestions was not clear to me.
